README.md
# Swedish Bank Account for PHP

[![Build Status](https://travis-ci.org/brocc-ab/se-bank-account-php.svg?branch=master)](https://travis-ci.org/brocc-ab/se-bank-account-php)
[![Latest Version](https://img.shields.io/github/release/brocc-ab/se-bank-account-php.svg?style=flat-square)](https://github.com/brocc-ab/se-bank-account-php/releases)
![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)

This package easily validates and returns details about Swedish bank account numbers using PHP. The details and validation rules are based on the documentation provided by Bankgirot, which can be found [here](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf).

## Installation

Installing the Swedish Bank Account for PHP can be done using Composer.

```bash
composer require brocc/se-bank-account
```

## Usage

Usage is very straight forward, just create a new instance of `Brocc\SeBankAccount\Account` which takes a string or an integer as argument and returns the details about the given number such as clearing number, account number and the name of the bank (if valid).

```php
<?php

require 'vendor/autoload.php';

use Brocc\SeBankAccount\Account;

// Instantiate using full account number incl. clearing number.
// Note: the number can be in any format, only digits are validated and used.
$account = new Account('3300-7505092556');

// Check if valid, returns true or false.
$account->isValid(); // true

// Returns the clearing number.
$account->clearing(); // 3300

// Returns the account number excl. clearing number, null if not valid.
$account->number(); // 7505092556

// Returns the name of the bank.
$account->bank(); // Nordea Personkonto
```

If an account is not valid `$account->isValid()` will return `false` and the other methods will return `null`.

## Supported Banks

Following Swedish banks are currently supported:

* Aion Bank SA
* Avanza Bank
* BNP Paribas SA
* Bankgirot
* BlueStep Finans
* Citibank
* DNB Bank
* Danske Bank
* Ekobanken
* Erik Penser
* Forex Bank *(discontinued\*)*
* Handelsbanken
* IKANO Bank
* Ica Banken
* JAK Medlemsbank
* Klarna Bank
* Landshypotek
* Lunar Bank A/S
* Länsförsäkringar Bank
* Lån & Spar Bank Sverige
* Marginalen Bank
* MedMera Bank *(discontinued\*)*
* Multitude Bank plc
* Nordax Bank *(NOBA Bank Group\*\*)*
* Nordea
* Nordea Personkonto
* Nordea/Plusgirot
* Nordnet Bank
* Northmill Bank
* Resurs Bank
* Riksgälden
* SBAB
* SEB
* Santander Consumer Bank
* Skandiabanken
* Sparbanken Syd
* Svea Bank
* Swedbank
* Swedbank fd. Sparbanken Öresund
* Ålandsbanken Sverige AB

\* Bank is currently not available in Bankgirot's documentation, available in package for backwards compatibility.\
\*\* Package will return Nordax Bank, even though the company name has changed to NOBA Bank Group.

## Contributing

If you find a bug or a bank that might not be supported, please submit an issue on Github directly under [issues](https://github.com/brocc-ab/se-bank-account-php/issues).

Any contributions are welcome!


CHANGELOG.md
### 1.1.0 (2023-11-10)
## Fixes
- [#2](https://github.com/brocc-ab/se-bank-account-php/issues/2): Bank list updated to align with the specification of Bankgirot.
- Fixed incorrect Nordea account number length validation reported in https://github.com/brocc-ab/se-bank-account-js/issues/4.


### 1.0.0 (2020-09-23)

- Initial release


tests/UtilsTest.php
<?php

/*
 * This file is part of Brocc\SeBankAccount.
 *
 * (c) Brocc AB <tech@brocc.se>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Brocc\SeBankAccount\Tests;

use Brocc\SeBankAccount\Utils;
use PHPUnit\Framework\TestCase;

class UtilsTest extends TestCase
{
    /**
     * @dataProvider mod10TestCases
     */
    public function testMod10Method(string $input, bool $valid)
    {
        $this->assertEquals($valid, Utils::mod10($input));
    }

    /**
     * @dataProvider mod11TestCases
     */
    public function testMod11Method(string $input, bool $valid)
    {
        $this->assertEquals($valid, Utils::mod11($input));
    }

    public function mod10TestCases(): array
    {
        return [
            [
                # Example from Bankgirot.
                'input' => '3316812057492',
                'valid' => true,
            ],
            [
                # Example from Bankgirot.
                'input' => '55555551',
                'valid' => true,
            ],
            [
                # Example from Bankgirot.
                'input' => '9912346',
                'valid' => true,
            ],
            [
                'input' => '3316812057490',
                'valid' => false,
            ],
            [
                'input' => '55555550',
                'valid' => false,
            ],
            [
                'input' => '9912340',
                'valid' => false,
            ],
            [
                # Swedish ssn.
                'input' => '9906130241',
                'valid' => true,
            ],
            [
                'input' => '9906130240',
                'valid' => false,
            ],
        ];
    }

    public function mod11TestCases(): array
    {
        return [
            [
                # Example from Bankgirot.
                'input' => '1912763608957',
                'valid' => true,
            ],
            [
                # Example from Bankgirot.
                'input' => '241350',
                'valid' => true,
            ],
            [
                # Example from Bankgirot.
                'input' => '324558',
                'valid' => true,
            ],
            [
                'input' => '1912763608950',
                'valid' => false,
            ],
            [
                'input' => '241351',
                'valid' => false,
            ],
            [
                'input' => '324550',
                'valid' => false,
            ],
        ];
    }
}


tests/AccountTest.php
<?php

/*
 * This file is part of Brocc\SeBankAccount.
 *
 * (c) Brocc AB <tech@brocc.se>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Brocc\SeBankAccount\Tests;

use Brocc\SeBankAccount\Account;
use PHPUnit\Framework\TestCase;

class AccountTest extends TestCase
{
    /**
     * @dataProvider accounts
     */
    public function testValidBankAccountNumbers(string $bank, string $number)
    {
        $account = new Account($number);

        // Split clearing and account number.
        $parts = explode('-', $number);

        $this->assertTrue($account->isValid());
        $this->assertEquals($parts[0], $account->clearing());
        $this->assertEquals($parts[1], $account->number());
        $this->assertEquals($bank, $account->bank());
    }

    /**
     * @dataProvider invalidBankAccountNumbers
     */
    public function testInvalidBankAccountNumber(string $bankAccountNumber)
    {
        $account = new Account($bankAccountNumber);

        $this->assertFalse($account->isValid());
        $this->assertNull($account->clearing());
        $this->assertNull($account->number());
        $this->assertNull($account->bank());
    }

    public function accounts(): array
    {
        return [
            [
                'bank' => 'Svea Bank',
                'number' => '9660-1000010',
            ],
            [
                'bank' => 'Aion Bank SA',
                'number' => '9580-1000100',
            ],
            [
                'bank' => 'Avanza Bank',
                'number' => '9550-0100000',
            ],
            [
                'bank' => 'BlueStep Finans',
                'number' => '9680-0000000',
            ],
            [
                'bank' => 'BNP Paribas SA',
                'number' => '9470-1000010',
            ],
            [
                'bank' => 'Citibank',
                'number' => '9040-1000100',
            ],
            [
                'bank' => 'Danske Bank',
                'number' => '1200-0000010',
            ],
            [
                'bank' => 'DNB Bank',
                'number' => '9190-1000100',
            ],
            [
                'bank' => 'Ekobanken',
                'number' => '9700-1000010',
            ],
            [
                'bank' => 'Erik Penser',
                'number' => '9590-0000100',
            ],
            [
                'bank' => 'Forex Bank',
                'number' => '9400-0001000',
            ],
            [
                'bank' => 'Ica Banken',
                'number' => '9270-0010000',
            ],
            [
                'bank' => 'IKANO Bank',
                'number' => '9170-0001000',
            ],
            [
                'bank' => 'JAK Medlemsbank',
                'number' => '9670-0000000',
            ],
            [
                'bank' => 'Klarna Bank',
                'number' => '9780-0000100',
            ],
            [
                'bank' => 'Landshypotek',
                'number' => '9390-0000001',
            ],
            [
                'bank' => 'Lunar Bank A/S',
                'number' => '9710-0000000',
            ],
            [
                'bank' => 'Lån & Spar Bank Sverige',
                'number' => '9630-0000001',
            ],
            [
                'bank' => 'Länsförsäkringar Bank',
                'number' => '3400-0001000',
            ],
            [
                'bank' => 'Länsförsäkringar Bank',
                'number' => '9020-0100000',
            ],
            [
                'bank' => 'Marginalen Bank',
                'number' => '9230-1000001',
            ],
            [
                'bank' => 'MedMera Bank',
                'number' => '9650-1000000',
            ],
            [
                'bank' => 'Multitude Bank plc',
                'number' => '9070-0000100',
            ],
            [
                'bank' => 'Nordax Bank',
                'number' => '9640-0010000',
            ],
            [
                'bank' => 'Nordea',
                'number' => '1100-0000001',
            ],
            [
                'bank' => 'Nordea',
                'number' => '1400-0001000',
            ],
            [
                'bank' => 'Nordea',
                'number' => '3000-1001000',
            ],
            [
                'bank' => 'Nordea',
                'number' => '3410-0100000',
            ],
            [
                'bank' => 'Nordea',
                'number' => '4000-1000000',
            ],
            [
                'bank' => 'Nordnet Bank',
                'number' => '9100-0000100',
            ],
            [
                'bank' => 'Northmill Bank',
                'number' => '9750-1000001',
            ],
            [
                'bank' => 'Resurs Bank',
                'number' => '9280-1000000',
            ],
            [
                'bank' => 'Riksgälden',
                'number' => '9880-0001000',
            ],
            [
                'bank' => 'Santander Consumer Bank',
                'number' => '9460-0010000',
            ],
            [
                'bank' => 'SBAB',
                'number' => '9250-0000001',
            ],
            [
                'bank' => 'SEB',
                'number' => '5000-1001000',
            ],
            [
                'bank' => 'SEB',
                'number' => '9120-0010000',
            ],
            [
                'bank' => 'SEB',
                'number' => '9130-1000000',
            ],
            [
                'bank' => 'Skandiabanken',
                'number' => '9150-0000010',
            ],
            [
                'bank' => 'Swedbank',
                'number' => '7000-1001000',
            ],
            [
                'bank' => 'Ålandsbanken Sverige AB',
                'number' => '2300-0000001',
            ],
            [
                'bank' => 'Danske Bank',
                'number' => '9180-1000000008',
            ],
            [
                'bank' => 'Handelsbanken',
                'number' => '6000-100000010',
            ],
            [
                'bank' => 'Nordea/Plusgirot',
                'number' => '9500-1000000008',
            ],
            [
                'bank' => 'Nordea Personkonto',
                'number' => '3300-7505092556',
            ],
            [
                'bank' => 'Nordea Personkonto',
                'number' => '3782-7505092556',
            ],
            [
                'bank' => 'Riksgälden',
                'number' => '9890-1000000008',
            ],
            [
                'bank' => 'Sparbanken Syd',
                'number' => '9570-1000000008',
            ],
            [
                'bank' => 'Swedbank',
                'number' => '80001-1000000008',
            ],
            [
                'bank' => 'Swedbank fd. Sparbanken Öresund',
                'number' => '9300-1000000008',
            ],
            [
                'bank' => 'Bankgirot',
                'number' => '9900-2000008',
            ],
        ];
    }

    public function invalidBankAccountNumbers(): array
    {
        return [
            ['0000-0000000'],
            ['1400-00010001234'],
            ['2000-10000040'],
            ['3301-010000099'],
            ['3410-0100000000000000000'],
            ['4000-10000003333']
        ];
    }
}


src/Utils.php
<?php

/*
 * This file is part of Brocc\SeBankAccount.
 *
 * (c) Brocc AB <tech@brocc.se>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Brocc\SeBankAccount;

class Utils
{
    /**
     * Mod-10 validation.
     *
     * @param string $number
     *
     * @return bool
     */
    public static function mod10($number)
    {
        $reversed = str_split(strrev($number));

        $sum = 0;
        foreach ($reversed as $i => $digit) {
            $product = 0 === (int) $i % 2 ? $digit : 2 * $digit;

            $sum += $product > 9 ? $product - 9 : $product;
        }

        return $sum && 0 === $sum % 10;
    }

    /**
     * Mod-11 validation.
     *
     * @param string $number
     *
     * @return bool
     */
    public static function mod11($number)
    {
        $reversed = str_split(strrev($number));

        $sum = 0;
        foreach ($reversed as $i => $digit) {
            $factor = ($i % 10) + 1;

            $sum += (int) $factor * $digit;
        }

        return $sum && 0 === $sum % 11;
    }
}


src/Account.php
<?php

/*
 * This file is part of Brocc\SeBankAccount.
 *
 * (c) Brocc AB <tech@brocc.se>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Brocc\SeBankAccount;

class Account
{
    /**
     * @var bool
     */
    private $isValid;

    /**
     * @var null|string
     */
    private $bank;

    /**
     * @var null|string
     */
    private $clearing;

    /**
     * @var null|string
     */
    private $number;

    /**
     * @param int|string $number
     */
    public function __construct($number)
    {
        $this->isValid = $this->init(preg_replace('/\D/i', '', $number));
    }

    /**
     * Returns the name of the bank.
     */
    public function bank(): ?string
    {
        return $this->bank;
    }

    /**
     * Returns the account clearing number, 4-5 digits.
     */
    public function clearing(): ?string
    {
        return $this->clearing;
    }

    /**
     * Returns the actual account number excl. clearing number.
     */
    public function number(): ?string
    {
        return $this->number;
    }

    /**
     * Check if given account number is valid.
     */
    public function isValid(): bool
    {
        return $this->isValid;
    }

    /**
     * Validates account number against all available banks.
     */
    private function init(string $number): bool
    {
        $banks = require __DIR__.'/banks.php';

        foreach ($banks as $bank) {
            if (preg_match($bank['regex'], $number)) {
                return $this->validate($number, $bank);
            }
        }

        return false;
    }

    /**
     * Validates account number against one bank.
     */
    private function validate(string $number, array $bank): bool
    {
        $clen = $bank['clen'] ?? 4;

        $clearing = substr($number, 0, $clen);
        $account = substr($number, $clen);

        // Only check type 1 with comment 1 & 2, for remaining combinations full account number excl. clearing is checked.
        $digits = 1 === $bank['type'] && 1 === $bank['comment']
            ? substr($clearing, 1).$account
            : (1 === $bank['type'] && 2 === $bank['comment'] ? $clearing.$account : $account);

        $valid = isset($bank['mod10']) && true === $bank['mod10'] ? Utils::mod10($digits) : Utils::mod11($digits);

        if (!$valid) {
            return false;
        }

        $this->bank = $bank['name'];
        $this->clearing = $clearing;
        $this->number = $account;

        return true;
    }
}


src/banks.php
<?php

/*
 * This file is part of Brocc\SeBankAccount.
 *
 * (c) Brocc AB <tech@brocc.se>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

return [
    [
        'name' => 'Svea Bank',
        'regex' => '/^966[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Aion Bank SA',
        'regex' => '/^958[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Avanza Bank',
        'regex' => '/^95[5-6][0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'BlueStep Finans',
        'regex' => '/^968[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'BNP Paribas SA',
        'regex' => '/^947[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Citibank',
        'regex' => '/^904[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Danske Bank',
        'regex' => '/^1[2-3][0-9][0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Danske Bank',
        'regex' => '/^24[0-9][0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'DNB Bank',
        'regex' => '/^(919|926)[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Ekobanken',
        'regex' => '/^970[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Erik Penser',
        'regex' => '/^959[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Forex Bank',
        'regex' => '/^94[0-4][0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Ica Banken',
        'regex' => '/^927[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'IKANO Bank',
        'regex' => '/^917[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'JAK Medlemsbank',
        'regex' => '/^967[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Klarna Bank',
        'regex' => '/^978[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Landshypotek',
        'regex' => '/^939[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Lunar Bank A/S',
        'regex' => '/^971[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Lån & Spar Bank Sverige',
        'regex' => '/^963[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Länsförsäkringar Bank',
        'regex' => '/^(340|906)[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Länsförsäkringar Bank',
        'regex' => '/^902[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Marginalen Bank',
        'regex' => '/^923[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'MedMera Bank',
        'regex' => '/^965[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Multitude Bank plc',
        'regex' => '/^907[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Nordax Bank',
        'regex' => '/^964[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Nordea',
        'regex' => '/^11[0-9]{9}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Nordea',
        'regex' => '/^(1[4-9][0-9]{9}$|20[0-9]{9}$)/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Nordea',
        'regex' => '/^(?!3300)(3[0-3][0-9]{9}$)/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Nordea',
        'regex' => '/^(?!3782)(3[4-9][1-9][0-9]{8}$)/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Nordea',
        'regex' => '/^4[0-9]{10}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Nordnet Bank',
        'regex' => '/^910[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Northmill Bank',
        'regex' => '/^975[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Resurs Bank',
        'regex' => '/^928[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Riksgälden',
        'regex' => '/^988[0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Santander Consumer Bank',
        'regex' => '/^946[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'SBAB',
        'regex' => '/^925[0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'SEB',
        'regex' => '/^5[0-9]{10}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'SEB',
        'regex' => '/^912[0-4][0-9]{7}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'SEB',
        'regex' => '/^91[3-4][0-9]{8}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Skandiabanken',
        'regex' => '/^91[5-6][0-9]{8}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Swedbank',
        'regex' => '/^7[0-9]{10}$/',
        'type' => 1,
        'comment' => 1,
    ],
    [
        'name' => 'Ålandsbanken Sverige AB',
        'regex' => '/^23[0-9]{9}$/',
        'type' => 1,
        'comment' => 2,
    ],
    [
        'name' => 'Danske Bank',
        'regex' => '/^918[0-9]{11}$/',
        'type' => 2,
        'comment' => 1,
        'mod10' => true,
    ],
    [
        'name' => 'Handelsbanken',
        'regex' => '/^6[0-9]{10,12}$/',
        'type' => 2,
        'comment' => 2,
    ],
    [
        'name' => 'Nordea/Plusgirot',
        'regex' => '/^(95[0-4]|996)[0-9]{8,11}$/',
        'type' => 2,
        'comment' => 3,
        'mod10' => true,
    ],
    [
        'name' => 'Nordea Personkonto',
        'regex' => '/^(3300|3782)[0-9]{10}$/',
        'type' => 2,
        'comment' => 1,
        'mod10' => true,
    ],
    [
        'name' => 'Riksgälden',
        'regex' => '/^989[0-9]{11}$/',
        'type' => 2,
        'comment' => 1,
        'mod10' => true,
    ],
    [
        'name' => 'Sparbanken Syd',
        'regex' => '/^957[0-9]{11}$/',
        'type' => 2,
        'comment' => 1,
        'mod10' => true,
    ],
    [
        'name' => 'Swedbank',
        'regex' => '/^8[0-9]{10,14}$/',
        'type' => 2,
        'comment' => 3,
        'mod10' => true,
        'clen' => 5,
    ],
    [
        'name' => 'Swedbank fd. Sparbanken Öresund',
        'regex' => '/^93[0-4][0-9]{11}$/',
        'type' => 2,
        'comment' => 1,
        'mod10' => true,
    ],
    [
        // Note: Bankgirot is not included in their own documentation for Swedish bank account numbers .
        'name' => 'Bankgirot',
        'regex' => '/^9900[0-9]{7,8}$/',
        'type' => null,
        'comment' => null,
        'mod10' => true,
    ],
];
